### PR TITLE
LISAMZA-36419: support offspring 

### PR DESCRIPTION
**Pull request by MabelYC**
**Please** add a meaningful description for your change here
LISAMZA-36419: support offspring 

initialize offspring generator of `UserPipelineOptionsFactory.class` when we call pipelineOptions.as(UserPipelineOptionsFactory.class). The Factory is singleton and also beam provides a cache of known 
pipelineOptions so we don't need to care about the factory will be initialized several times.

Tested locally with snapshot versions.
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/InvokePipelineOptionsFactory.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.Iterator;
+import java.util.ServiceLoader;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterators;
+import org.checkerframework.checker.initialization.qual.Initialized;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+
+@SuppressWarnings("rawtypes")
+/**
+ * Interface to support offspring wire-in for Li
```
Comment (yananhao12) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/InvokePipelineOptionsFactory.java]: Can you elaborate more on how this interface is supposed to be used, as the term `InvokePipelineOptionsFactory` seems ambiguous to me.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/ProxyInvocationHandler.java [hunk 2]
```diff
@@ -275,6 +275,10 @@ static BoundValue fromDefault(@Nullable Object value) {
    * @return An object that implements the interface {@code <T>}.
    */
   <T extends PipelineOptions> T as(Class<T> iface) {
+    return as(iface, null);
+  }
+
+  <T extends PipelineOptions> T as(Class<T> iface, PipelineOptions pipelineOptions) {
```
Comment (yananhao12) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/ProxyInvocationHandler.java:281]: Is the only purpose of passing `pipelineOptions` to tell whether `InvokePipelineOptionsFactory` should be called? What are the cases when `pipelienOptions` is null?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/InvokePipelineOptionsFactory.java (hunk 1 above)
Comment (MabelYC) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/InvokePipelineOptionsFactory.java]: Add some comments.
So this interface is to designed to check if the input class meets the requirements to initialize some other object underneath to support further execution.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/ProxyInvocationHandler.java (hunk 2 above)
Comment (MabelYC) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/ProxyInvocationHandler.java:281]: there're two place 1. calling xxPipelineOptions.as(pipelineOptions.class) to return a another type of pipelineOptions. 2. creating the a raw pipelineOptions using PipelineOptionsFactory.create(). 
#1 is where we want to initialize the generator.
and when execute #2 initialize the generator is already wired in.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+@SuppressWarnings("rawtypes")
+public interface CustomPipelineOptionsInitializer<T> {
+  void init(T pipelineOptions, Class<T> clazz);
+
+  interface Registrar {
+    CustomPipelineOptionsInitializer create();
+  }
+
+
+  static @Initialized @Nullable CustomPipelineOptionsInitializer getFactory() {
```
Comment (yananhao12) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java]: should we just call this method get()?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+/**
+ * Interface to support offspring wire-in for Li: if input class meets some requirements,
+ * the customized pipelineOptions will be initialized.
+ *
+ */
+@SuppressWarnings("rawtypes")
+public interface CustomPipelineOptionsInitializer<T> {
+  void init(T pipelineOptions, Class<T> clazz);
+
+  interface Registrar {
```
Comment (yananhao12) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java:33]: how will the Registrar be used?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+import java.util.Iterator;
+import java.util.ServiceLoader;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterators;
+import org.checkerframework.checker.initialization.qual.Initialized;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+
+/**
+ * Interface to support offspring wire-in for Li: if input class meets some requirements,
```
Comment (yananhao12) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java]: `meets some requires` doesn't convey any useful information here.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java (hunk 4 above)
Comment (MabelYC) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java:33]: In the implementation, we will have a class annotated as AutoService(CustomePipelineOptionsInitializer.Registrar.class). It returns an instance of the implementation, so beam will know which implementation to use during runtime

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java (hunk 5 above)
Comment (MabelYC) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java]: Updated, please take a look.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  void init(T pipelineOptions, Class<T> clazz);
+
+  interface Registrar {
+    CustomPipelineOptionsInitializer create();
+  }
+
+
+  static @Initialized @Nullable CustomPipelineOptionsInitializer get() {
+    final Iterator<CustomPipelineOptionsInitializer.Registrar>
+        factories = ServiceLoader.load(CustomPipelineOptionsInitializer.Registrar.class).iterator();
```
Comment (yananhao12) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java]: should the variable name be renamed `initializers`?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java (hunk 4 above)
Comment (yananhao12) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java:33]: Got it. Can you add an example of usage in the javadoc?

Comment (MabelYC) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/options/CustomPipelineOptionsInitializer.java:33]: Updated

**Review by yananhao12 — APPROVED** [review]

**Pull request merged**